### PR TITLE
chore(workspace): Migrate back to `thiserror` v2

### LINKED ISSUE #806: chore: Replace Custom error impls with thiserror
### Description

Since `thiserror` [v2 was released](https://github.com/dtolnay/thiserror), `kona` can now use the `no_std` compatible `thiserror` for use in defining custom error types instead of manually implementing the `core::error::Error` trait for error types with `derive_more::Display`. `From<some_type> for CustomError` should also be replaced with the `#[from]` syntax sugar provided by thiserror.

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Migrates all crates back to `thiserror`, now that `>=v2.0.0` has support for `no_std` with `default-features = false`.

closes #806

### CODE REVIEW

**Comment by clabby** [comment]
## 📚 $\text{Stack Overview}$

Pulls submitted in this stack:
* #811 👈
* `main`

_This comment was automatically generated by [`st`](https://github.com/clabby/st)._

File: crates/derive/src/errors/pipeline.rs [hunk 1]
```diff
@@ -17,17 +18,17 @@ macro_rules! ensure {
...
     /// A critical error.
-    #[display("Critical error: {_0}")]
-    Critical(PipelineError),
+    #[error("Critical error: {0}")]
+    Critical(#[source] PipelineError),
     /// A reset error.
-    #[display("Pipeline reset: {_0}")]
-    Reset(ResetError),
+    #[error("Pipeline reset: {0}")]
+    Reset(#[source] ResetError),
```
Comment (refcell) [inline crates/derive/src/errors/pipeline.rs:31]: Can we also remove the `impl From<ResetError>` below and inline with a `#[from]`?

Comment (clabby) [reply crates/derive/src/errors/pipeline.rs]: Yep! Done

**Review by refcell — APPROVED** [review]
Very nice mr allen

**Pull request merged**